### ISSUE TITLE
Inconsistent type inference on hover vs autocompletion

### ISSUE THREAD

**Issue by gaurav5430** [microsoft/TypeScript#45140]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.58.2 (was happening before this version as well)
- OS Version: Mac OS 10.14

Steps to Reproduce:

1. Create a React functional component like this:

```
import React from "react";
import PropTypes from "prop-types";

//  vscode is able to infer the types correctly in autocomplete
// as this is a functional component, with prop types
// Please note that vscode still does not show the correct types
// when hovering on the component usage

function MyComponent({ prop1, prop2, prop3 }) {
  return (
    <div>
      <span>{prop1}</span>
      <span>{prop2}</span>
    </div>
  );
}

MyComponent.propTypes = {
  prop1: PropTypes.string.isRequired,
  prop2: PropTypes.bool,
  prop3: PropTypes.shape({
    p1: PropTypes.string
  })
};

MyComponent.defaultProps = {
  prop2: false
};

export default MyComponent;
```

2. At the point of usage, where you have imported this component, hover on the usage, you will see that vscode is able to identify the function signature, but not the types of the function arguments, and shows them as `any`

<img width="697" alt="Screenshot 2021-07-21 at 1 30 58 AM" src="https://user-images.githubusercontent.com/12510548/126449768-94081dd0-f5df-4cb7-a2c7-a1742c13f96c.png">

3. If you hovber over the specific prop though, you would see that vscode shows the correct type (inferred from prop types) and also provides autocompletion for the prop name. (Although there is still an issue that it shows all the props as optional)

<img width="1034" alt="Screenshot 2021-07-21 at 1 31 04 AM" src="https://user-images.githubusercontent.com/12510548/126449980-beadb6be-b885-4c80-8db6-78078f65fe9e.png">

### Example:
https://codesandbox.io/s/vscode-intellisense-proptypes-fc-w-destructuring-w-proptypes-cngth?from-embed=&file=/src/MyComponent.js:0-655

### Expectation:
I would have expected that 
- vscode would be able to infer the types on hovering as well, same as it can do while providing autocomplete suggestion for the prop
- vscode would not show all the props as optional, since it seems that it can infer prop names and types from prop types, it should also be able to infer whether it is required or optional.

**Comment by andrewbranch** [comment]
What’s being shown is correct. The types of `prop1`, `prop2`, and `prop3` at/in the `MyComponent` declaration _are_ all `any`, and that’s what quick info is showing. However, JSX support in TypeScript allows JSX environments (like React) to define a type called `LibraryManagedAttributes` which takes the type of the component being created in JSX and declared type of the props, and it can return a new type to be used for the type of the props that you actually pass in JSX, which is how React is able to use the `.propTypes` to infer actual types for JSX props. However, that has _no effect_ on the props _outside_ of JSX element creation. You can hover `prop1` on the JSX element itself and it shows the JSX props type. Hovering `MyComponent` simply shows the parameter type of the `MyComponent` function, separate from any augmentation via JSX.

**Comment by gaurav5430** [comment]
ok

so:
- there is no easy way for vscode to use the `LibraryManagedAttributes` in quick info ?
- the props are shown as optional because of how the JSX environment infers the prop types, and that cannot be fixed? if yes, can raise a issue for this in appropriate repo, if you can point me to it.

**Comment by andrewbranch** [comment]
1. We could choose to make some sort of JSX-specific quick info while hovering JSX element components, but would need a proposal for what that should look like.
2. Not sure what the problem is with regard to optional props; looks like completions correctly shows `prop1` as required and `prop2` and `prop3` as optional, which is what the PropTypes info says.

**Comment by gaurav5430** [comment]
screenshot showing prop1 is required but also optional 

<img width="675" alt="Screenshot 2021-07-28 at 11 59 52 PM" src="https://user-images.githubusercontent.com/12510548/127376795-d74717ee-9ae6-4d7a-b669-1ccb63434418.png">

**Comment by andrewbranch** [comment]
That seems to be some sort of artifact of our looser JS checking for functions with untyped parameters. It might be something we can improve, but note that all these problems go away if you just add proper JSDoc parameter annotations on `MyComponent`.

**Comment by gaurav5430** [comment]
yeah, we have been using jsdoc comments in some cases as well, but that just adds up one more point for documentation. We use proptypes for dynamic type checking, if we add jsdoc comments as well, that slightly duplicates the effort. Some of the tools that we use can read comments on the prop types and generate documentation, so we add comments there anyway, we would need to repeat that in jsdoc

**Comment by andrewbranch** [comment]
I’ve moved that to #45223
